File: backend/error_handler.py
"""
Standardized Error Handling System for Eunoia Journal
Provides consistent error handling, logging, and response formatting across all services
"""

import logging
import traceback
import uuid
from datetime import datetime
from typing import Dict, Any, Optional, Union
from enum import Enum
from dataclasses import dataclass
from fastapi import HTTPException, status
from fastapi.responses import JSONResponse

# Set up structured logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

class ErrorCode(Enum):
    """Standardized error codes for the application"""
    # Authentication & Authorization
    AUTH_TOKEN_MISSING = "AUTH_001"
    AUTH_TOKEN_INVALID = "AUTH_002"
    AUTH_TOKEN_EXPIRED = "AUTH_003"
    AUTH_INSUFFICIENT_PERMISSIONS = "AUTH_004"
    
    # Database & Data
    DB_CONNECTION_ERROR = "DB_001"
    DB_QUERY_ERROR = "DB_002"
    DB_CONSTRAINT_ERROR = "DB_003"
    DATA_NOT_FOUND = "DATA_001"
    DATA_VALIDATION_ERROR = "DATA_002"
    DATA_DUPLICATE_ERROR = "DATA_003"
    
    # ML & AI Services
    ML_MODEL_LOAD_ERROR = "ML_001"
    ML_ANALYSIS_ERROR = "ML_002"
    ML_API_ERROR = "ML_003"
    ML_FALLBACK_ERROR = "ML_004"
    
    # External Services
    EXTERNAL_API_ERROR = "EXT_001"
    EXTERNAL_TIMEOUT = "EXT_002"
    EXTERNAL_RATE_LIMIT = "EXT_003"
    
    # General Application
    INTERNAL_SERVER_ERROR = "APP_001"
    VALIDATION_ERROR = "APP_002"
    CONFIGURATION_ERROR = "APP_003"
    NETWORK_ERROR = "APP_004"

class ErrorSeverity(Enum):
    """Error severity levels"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

@dataclass
class ErrorContext:
    """Context information for errors"""
    request_id: str
    user_id: Optional[str] = None
    endpoint: Optional[str] = None
    timestamp: datetime = None
    additional_data: Optional[Dict[str, Any]] = None
    
    def __post_init__(self):
        if self.timestamp is None:
            self.timestamp = datetime.utcnow()

@dataclass
class StandardError:
    """Standardized error structure"""
    code: ErrorCode
    message: str
    detail: Optional[str] = None
    severity: ErrorSeverity = ErrorSeverity.MEDIUM
    context: Optional[ErrorContext] = None
    original_exception: Optional[Exception] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert error to dictionary for JSON response"""
        return {
            "error": {
                "code": self.code.value,
                "message": self.message,
                "detail": self.detail,
                "severity": self.severity.value,
                "request_id": self.context.request_id if self.context else None,
                "timestamp": self.context.timestamp.isoformat() if self.context else None,
                "user_id": self.context.user_id if self.context else None,
                "endpoint": self.context.endpoint if self.context else None
            }
        }

class ErrorHandler:
    """Centralized error handling class"""
    
    @staticmethod
    def create_error_context(
        user_id: Optional[str] = None,
        endpoint: Optional[str] = None,
        additional_data: Optional[Dict[str, Any]] = None
    ) -> ErrorContext:
        """Create error context with unique request ID"""
        return ErrorContext(
            request_id=str(uuid.uuid4()),
            user_id=user_id,
            endpoint=endpoint,
            additional_data=additional_data
        )
    
    @staticmethod
    def log_error(error: StandardError, include_traceback: bool = True) -> None:
        """Log error with appropriate level based on severity"""
        log_message = f"[{error.code.value}] {error.message}"
        if error.detail:
            log_message += f" - {error.detail}"
        
        if error.context:
            log_message += f" | Request: {error.context.request_id}"
            if error.context.user_id:
                log_message += f" | User: {error.context.user_id}"
        
        if include_traceback and error.original_exception:
            log_message += f"\nTraceback: {traceback.format_exc()}"
        
        # Log based on severity
        if error.severity == ErrorSeverity.CRITICAL:
            logger.critical(log_message)
        elif error.severity == ErrorSeverity.HIGH:
            logger.error(log_message)
        elif error.severity == ErrorSeverity.MEDIUM:
            logger.warning(log_message)
        else:
            logger.info(log_message)
    
    @staticmethod
    def create_http_exception(
        error: StandardError,
        status_code: int = status.HTTP_500_INTERNAL_SERVER_ERROR
    ) -> HTTPException:
        """Create HTTPException from StandardError"""
        ErrorHandler.log_error(error)
        return HTTPException(
            status_code=status_code,
            detail=error.to_dict()
        )
    
    @staticmethod
    def create_json_response(
        error: StandardError,
        status_code: int = status.HTTP_500_INTERNAL_SERVER_ERROR
    ) -> JSONResponse:
        """Create JSONResponse from StandardError"""
        ErrorHandler.log_error(error)
        return JSONResponse(
            status_code=status_code,
            content=error.to_dict()
        )

# Predefined error creators for common scenarios
class ErrorFactory:
    """Factory for creating common error types"""
    
    @staticmethod
    def authentication_error(
        message: str = "Authentication failed",
        detail: Optional[str] = None,
        context: Optional[ErrorContext] = None
    ) -> StandardError:
        """Create authentication error"""
        return StandardError(
            code=ErrorCode.AUTH_TOKEN_INVALID,
            message=message,
            detail=detail,
            severity=ErrorSeverity.HIGH,
            context=context
        )
    
    @staticmethod
    def validation_error(
        message: str = "Validation failed",
        detail: Optional[str] = None,
        context: Optional[ErrorContext] = None
    ) -> StandardError:
        """Create validation error"""
        return StandardError(
            code=ErrorCode.VALIDATION_ERROR,
            message=message,
            detail=detail,
            severity=ErrorSeverity.MEDIUM,
            context=context
        )
    
    @staticmethod
    def database_error(
        message: str = "Database operation failed",
        detail: Optional[str] = None,
        context: Optional[ErrorContext] = None,
        original_exception: Optional[Exception] = None
    ) -> StandardError:
        """Create database error"""
        return StandardError(
            code=ErrorCode.DB_QUERY_ERROR,
            message=message,
            detail=detail,
            severity=ErrorSeverity.HIGH,
            context=context,
            original_exception=original_exception
        )
    
    @staticmethod
    def ml_service_error(
        message: str = "ML analysis failed",
        detail: Optional[str] = None,
        context: Optional[ErrorContext] = None,
        original_exception: Optional[Exception] = None
    ) -> StandardError:
        """Create ML service error"""
        return StandardError(
            code=ErrorCode.ML_ANALYSIS_ERROR,
            message=message,
            detail=detail,
            severity=ErrorSeverity.MEDIUM,
            context=context,
            original_exception=original_exception
        )
    
    @staticmethod
    def external_service_error(
        message: str = "External service error",
        detail: Optional[str] = None,
        context: Optional[ErrorContext] = None,
        original_exception: Optional[Exception] = None
    ) -> StandardError:
        """Create external service error"""
        return StandardError(
            code=ErrorCode.EXTERNAL_API_ERROR,
            message=message,
            detail=detail,
            severity=ErrorSeverity.MEDIUM,
            context=context,
            original_exception=original_exception
        )
    
    @staticmethod
    def not_found_error(
        resource: str = "Resource",
        context: Optional[ErrorContext] = None
    ) -> StandardError:
        """Create not found error"""
        return StandardError(
            code=ErrorCode.DATA_NOT_FOUND,
            message=f"{resource} not found",
            detail=f"The requested {resource.lower()} could not be found",
            severity=ErrorSeverity.MEDIUM,
            context=context
        )
    
    @staticmethod
    def internal_server_error(
        message: str = "Internal server error",
        detail: Optional[str] = None,
        context: Optional[ErrorContext] = None,
        original_exception: Optional[Exception] = None
    ) -> StandardError:
        """Create internal server error"""
        return StandardError(
            code=ErrorCode.INTERNAL_SERVER_ERROR,
            message=message,
            detail=detail,
            severity=ErrorSeverity.CRITICAL,
            context=context,
            original_exception=original_exception
        )

# Decorator for automatic error handling
def handle_errors(
    error_code: ErrorCode = ErrorCode.INTERNAL_SERVER_ERROR,
    severity: ErrorSeverity = ErrorSeverity.MEDIUM,
    include_traceback: bool = True
):
    """Decorator for automatic error handling in service methods"""
    def decorator(func):
        def wrapper(*args, **kwargs):
            try:
                return func(*args, **kwargs)
            except Exception as e:
                context = ErrorHandler.create_error_context()
                error = StandardError(
                    code=error_code,
                    message=f"Error in {func.__name__}: {str(e)}",
                    detail=str(e),
                    severity=severity,
                    context=context,
                    original_exception=e
                )
                ErrorHandler.log_error(error, include_traceback)
                raise ErrorHandler.create_http_exception(error)
        return wrapper
    return decorator

# Global error handler instance
error_handler = ErrorHandler()
error_factory = ErrorFactory()

File: frontend/src/utils/errorHandler.ts
/**
 * Standardized Error Handling System for Eunoia Journal Frontend
 * Provides consistent error handling, logging, and user feedback across all components
 */

export enum ErrorCode {
  // Network & API Errors
  NETWORK_ERROR = 'NET_001',
  API_TIMEOUT = 'NET_002',
  API_UNAUTHORIZED = 'NET_003',
  API_FORBIDDEN = 'NET_004',
  API_NOT_FOUND = 'NET_005',
  API_SERVER_ERROR = 'NET_006',

  // Validation Errors
  VALIDATION_ERROR = 'VAL_001',
  REQUIRED_FIELD_MISSING = 'VAL_002',
  INVALID_FORMAT = 'VAL_003',

  // Authentication Errors
  AUTH_TOKEN_MISSING = 'AUTH_001',
  AUTH_TOKEN_EXPIRED = 'AUTH_002',
  AUTH_LOGIN_FAILED = 'AUTH_003',

  // Data Errors
  DATA_LOAD_ERROR = 'DATA_001',
  DATA_SAVE_ERROR = 'DATA_002',
  DATA_NOT_FOUND = 'DATA_003',

  // ML/AI Errors
  ML_ANALYSIS_ERROR = 'ML_001',
  ML_SERVICE_UNAVAILABLE = 'ML_002',

  // General Application Errors
  UNKNOWN_ERROR = 'APP_001',
  CONFIGURATION_ERROR = 'APP_002',
}

export enum ErrorSeverity {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

export interface ErrorContext {
  requestId: string;
  userId?: string;
  component?: string;
  action?: string;
  timestamp: Date;
  additionalData?: Record<string, unknown>;
}

export interface StandardError {
  code: ErrorCode;
  message: string;
  detail?: string;
  severity: ErrorSeverity;
  context: ErrorContext;
  originalError?: Error;
  userMessage?: string; // User-friendly message for display
}

export class ErrorHandler {
  private static instance: ErrorHandler;
  private errorLog: StandardError[] = [];

  private constructor() {
    // Private constructor for singleton pattern
  }

  static getInstance(): ErrorHandler {
    if (!ErrorHandler.instance) {
      ErrorHandler.instance = new ErrorHandler();
    }
    return ErrorHandler.instance;
  }

  /**
   * Create error context with unique request ID
   */
  createErrorContext(
    userId?: string,
    component?: string,
    action?: string,
    additionalData?: Record<string, unknown>,
  ): ErrorContext {
    return {
      requestId: this.generateRequestId(),
      userId,
      component,
      action,
      timestamp: new Date(),
      additionalData,
    };
  }

  /**
   * Generate unique request ID
   */
  private generateRequestId(): string {
    return `req_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Create standardized error from various error sources
   */
  createError(
    code: ErrorCode,
    message: string,
    options: {
      detail?: string;
      severity?: ErrorSeverity;
      context?: ErrorContext;
      originalError?: Error;
      userMessage?: string;
    } = {},
  ): StandardError {
    const {
      detail,
      severity = ErrorSeverity.MEDIUM,
      context,
      originalError,
      userMessage,
    } = options;

    const error: StandardError = {
      code,
      message,
      detail,
      severity,
      context: context || this.createErrorContext(),
      originalError,
      userMessage: userMessage || this.getUserFriendlyMessage(code, message),
    };

    this.logError(error);
    return error;
  }

  /**
   * Log error with appropriate level
   */
  logError(error: StandardError): void {
    // Add to error log
    this.errorLog.push(error);

    // Console logging based on severity
    const logMessage = `[${error.code}] ${error.message} | Request: ${error.context.requestId}`;

    switch (error.severity) {
      case ErrorSeverity.CRITICAL:
        // eslint-disable-next-line no-console
        console.error(logMessage, error);
        break;
      case ErrorSeverity.HIGH:
        // eslint-disable-next-line no-console
        console.error(logMessage, error);
        break;
      case ErrorSeverity.MEDIUM:
        // eslint-disable-next-line no-console
        console.warn(logMessage, error);
        break;
      case ErrorSeverity.LOW:
        // eslint-disable-next-line no-console
        console.info(logMessage, error);
        break;
    }

    // In production, you might want to send errors to a logging service
    if (process.env.NODE_ENV === 'production') {
      this.sendToLoggingService(error);
    }
  }

  /**
   * Get user-friendly error message
   */
  private getUserFriendlyMessage(code: ErrorCode, _message: string): string {
    const userMessages: Record<ErrorCode, string> = {
      [ErrorCode.NETWORK_ERROR]:
        'Unable to connect to the server. Please check your internet connection.',
      [ErrorCode.API_TIMEOUT]: 'The request is taking too long. Please try again.',
      [ErrorCode.API_UNAUTHORIZED]: 'Please log in to continue.',
      [ErrorCode.API_FORBIDDEN]: 'You do not have permission to perform this action.',
      [ErrorCode.API_NOT_FOUND]: 'The requested resource was not found.',
      [ErrorCode.API_SERVER_ERROR]: 'Something went wrong on our end. Please try again later.',
      [ErrorCode.VALIDATION_ERROR]: 'Please check your input and try again.',
      [ErrorCode.REQUIRED_FIELD_MISSING]: 'Please fill in all required fields.',
      [ErrorCode.INVALID_FORMAT]: 'Please check the format of your input.',
      [ErrorCode.AUTH_TOKEN_MISSING]: 'Please log in to continue.',
      [ErrorCode.AUTH_TOKEN_EXPIRED]: 'Your session has expired. Please log in again.',
      [ErrorCode.AUTH_LOGIN_FAILED]: 'Login failed. Please check your credentials.',
      [ErrorCode.DATA_LOAD_ERROR]: 'Failed to load data. Please try refreshing the page.',
      [ErrorCode.DATA_SAVE_ERROR]: 'Failed to save data. Please try again.',
      [ErrorCode.DATA_NOT_FOUND]: 'No data found.',
      [ErrorCode.ML_ANALYSIS_ERROR]:
        'AI analysis is temporarily unavailable. Your entry has been saved.',
      [ErrorCode.ML_SERVICE_UNAVAILABLE]: 'AI features are temporarily unavailable.',
      [ErrorCode.UNKNOWN_ERROR]: 'An unexpected error occurred. Please try again.',
      [ErrorCode.CONFIGURATION_ERROR]: 'Application configuration error. Please contact support.',
    };

    return userMessages[code] || 'An unexpected error occurred. Please try again.';
  }

  /**
   * Send error to logging service (placeholder for production)
   */
  private sendToLoggingService(error: StandardError): void {
    // In production, implement actual logging service integration
    // e.g., Sentry, LogRocket, etc.
    // eslint-disable-next-line no-console
    console.log('Sending error to logging service:', error);
  }

  /**
   * Handle API errors from axios responses
   */
  handleApiError(error: unknown, context?: Partial<ErrorContext>): StandardError {
    const errorContext = this.createErrorContext(
      context?.userId,
      context?.component,
      context?.action,
      context?.additionalData,
    );

    // Narrow unknown - check if it's an axios error
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const err = error as any;

    if (err.response) {
      // Server responded with error status
      const status = err.response.status;
      const data = err.response.data;

      let code: ErrorCode;
      let message: string;
      let severity: ErrorSeverity = ErrorSeverity.MEDIUM;

      switch (status) {
        case 400:
          code = ErrorCode.VALIDATION_ERROR;
          message = data?.detail || 'Bad request';
          break;
        case 401:
          code = ErrorCode.API_UNAUTHORIZED;
          message = 'Unauthorized';
          severity = ErrorSeverity.HIGH;
          break;
        case 403:
          code = ErrorCode.API_FORBIDDEN;
          message = 'Forbidden';
          severity = ErrorSeverity.HIGH;
          break;
        case 404:
          code = ErrorCode.API_NOT_FOUND;
          message = 'Not found';
          break;
        case 408:
          code = ErrorCode.API_TIMEOUT;
          message = 'Request timeout';
          break;
        case 500:
        case 502:
        case 503:
        case 504:
          code = ErrorCode.API_SERVER_ERROR;
          message = 'Server error';
          severity = ErrorSeverity.HIGH;
          break;
        default:
          code = ErrorCode.API_SERVER_ERROR;
          message = `HTTP ${status} error`;
      }

      return this.createError(code, message, {
        detail: data?.detail || String(err.message || ''),
        severity,
        context: errorContext,
        originalError: err,
      });
    } else if (err.request) {
      // Network error
      return this.createError(ErrorCode.NETWORK_ERROR, 'Network error', {
        detail: String(err.message || ''),
        severity: ErrorSeverity.HIGH,
        context: errorContext,
        originalError: err,
      });
    } else {
      // Other error
      return this.createError(ErrorCode.UNKNOWN_ERROR, 'Unknown error', {
        detail: String(err?.message || ''),
        context: errorContext,
        originalError: err,
      });
    }
  }

  /**
   * Get recent errors for debugging
   */
  getRecentErrors(limit: number = 10): StandardError[] {
    return this.errorLog.slice(-limit);
  }

  /**
   * Clear error log
   */
  clearErrorLog(): void {
    this.errorLog = [];
  }
}

// Export singleton instance
export const errorHandler = ErrorHandler.getInstance();

// Export error factory functions for common scenarios
export const createApiError = (error: unknown, context?: Partial<ErrorContext>) =>
  errorHandler.handleApiError(error, context);

export const createValidationError = (message: string, context?: Partial<ErrorContext>) =>
  errorHandler.createError(ErrorCode.VALIDATION_ERROR, message, {
    context: context
      ? errorHandler.createErrorContext(
          context.userId,
          context.component,
          context.action,
          context.additionalData,
        )
      : undefined,
  });

export const createNetworkError = (message: string, context?: Partial<ErrorContext>) =>
  errorHandler.createError(ErrorCode.NETWORK_ERROR, message, {
    context: context
      ? errorHandler.createErrorContext(
          context.userId,
          context.component,
          context.action,
          context.additionalData,
        )
      : undefined,
    severity: ErrorSeverity.HIGH,
  });

export const createDataError = (message: string, context?: Partial<ErrorContext>) =>
  errorHandler.createError(ErrorCode.DATA_LOAD_ERROR, message, {
    context: context
      ? errorHandler.createErrorContext(
          context.userId,
          context.component,
          context.action,
          context.additionalData,
        )
      : undefined,
  });

export const createMLError = (message: string, context?: Partial<ErrorContext>) =>
  errorHandler.createError(ErrorCode.ML_ANALYSIS_ERROR, message, {
    context: context
      ? errorHandler.createErrorContext(
          context.userId,
          context.component,
          context.action,
          context.additionalData,
        )
      : undefined,
    severity: ErrorSeverity.LOW, // ML errors are often non-critical
  });
